manager/eris-mint/eris-mint.go

```diff
@@ -185,15 +185,14 @@ func (app *ErisMint) Commit() (res tmsp.Result) {
 	// sync the AppendTx cache
 	app.cache.Sync()
 
-	// if there were any txs in the block,
-	// reset the check cache to the new height
-	if app.nTxs > 0 {
-		log.WithFields(log.Fields{
-			"txs": app.nTxs,
-		}).Info("Reset checkCache")
-		app.checkCache = sm.NewBlockCache(app.state)
-	}
-	app.nTxs = 0
+	// Refresh checkCache even if there were no transactions. A CheckTx may have
```

Comment:
> CheckTx can be made more strongly aware of its inability to acknowledge the ordering and as such can be made smarter on explicit validations of ordering;
I'm not entirely sure what you are are saying in your second comment

I don't really understand what you are saying here. The ordering of any previous block's transaction are obviously fixed

When we commit we want to make sure future CheckTx runs have a fresh state, checking here for whether there is an empty block is a completely premature optimisation. I think it gains us very little and makes the code harder to understand.

The issue is that CheckTx can get run once and before being added to a proposal a recheck is forced (by a new block), the recheck should be running against the last committed state not the dirty state from the last CheckTx. Strict inequality wouldn't help here, I have a Tx with a correct nonce, but it is deemed invalid because it's been applied to the state twice. Actually it's valid against the current committed state.

Come to think of it, we should really be refreshing the checkCache every time CheckTx is called! I will remove this from here and make an analternative proposal we can discuss.